compiler/z/codegen/BinaryEvaluator.cpp

```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
-         int32_t popCnt = populationCount(value);
-         int32_t msBit = -1;
-         int32_t lsBit = -1;
+         int32_t tZeros = trailingZeroes(longConstValue);
+         int32_t lZeros = leadingZeroes(longConstValue);
+         int32_t tOnes  = trailingZeroes(~longConstValue);
+         int32_t lOnes  = leadingZeroes(~longConstValue);
+         int32_t popCnt = populationCount(longConstValue);
+
+         // the minimum value for the bit positions is 0, and lsBit can equal msBit iff popCnt = 1
+         // if popCnt == 0 then we cannot generate a RISBG instruction because RISBG preserves at least one bit position
+         int32_t msBit = 0;           
+         int32_t lsBit = 0;  
 
          if (comp->getOption(TR_TraceCG))
             {
             traceMsg(comp,"[%p] land => rotated-and-insert: tZeros %d, lZeros %d, popCnt %d\n", node, tZeros, lZeros, popCnt);
             traceMsg(comp,"          => rotated-and-insert: lOnes %d, tOnes %d\n", lOnes, tOnes);
             }
 
-         bool doIt = false;
+         bool doTransformation = false;
+
+         // We now try to detect if the lconst's bit pattern is such that we can use a single RISBG
+         // instead of multiple AND instructions
+
+         // The first block will catch all cases where the lConstValue is a series of contiguous ones 
+         // surrounded by zeros
+         // ex: 0001111000, 011110, 111111, 001111, 11100 etc...
+         // 
+         // The 'else if' block will consider all cases where we have a series of contiguous zeros 
+         // surrounded by ones
+         // ex: 1110000111, 1000001
+         //
+         // Note: if the lconst is 0x0 we cannot generate a RISBG instruction.
+         // The starting and ending bit positions are inclusive, so specifying (lsBit,msBit)=(0,0) will
+         // preserve the 0th bit in the firstChild, which would be incorrect.
+         // If such a case arises where we can figure out at compile time that the 
+         // constant's value is 0, then we must handle that case in the simplifier.
          if (popCnt == (64 - lZeros - tZeros))
             {
-            msBit = lZeros;
+            msBit = lZeros; 
             lsBit = 63 - tZeros;
-            if (firstChild->getReferenceCount()>1 || skipedConversion != NULL   // Better of with non-destructive RISBG as AND?
-                  || lZeros > 31 || tZeros > 31 || (lZeros > 0 && tZeros > 0))  // Better off with a single AND?
+            if (firstChild->getReferenceCount() > 1 // if we use an AND and refCount>1, then we will have to backup the value in another register beforehand, as AND is destructive
+                || lZeros > 31 || tZeros > 31 // this makes sure we are not better off using one of the following AND instructions: NILF,NILH,NILL, NIHF,NIHH, NIHL
+                || (lZeros > 0 && tZeros > 0)) // if our constant at minimum looks like 0111...110, then it's better to use RISBG as this case will require multiple ANDs
```

Comment:
This does not look correct. What if `lZeros == 1` and `tZeros == 60`. In this case you would carry out the transformation even though you could have generated `NILL` which is a 4-byte instruction as opposed to the 6-byte `RISBG`. This needs more careful consideration.

z/Architecture has all variants of AND immediate instructions for the 16-bit non-overlapping parts of a register. We should only be using `RISBG` in the cases where we cannot handle the AND with a single such 16-bit AND immediate operation. We need to have checks for this.

Reply:
So I don't thing that example would work. A value such as that looks like this in binary:
011100000...0. We would need to explicitly zero out everything except those three bits. As per the specification for the 6 AND immediate instructions in Z, 

>"the second operand is ANDed with bits of the first operand, and the result replaces those bits of the first operand. The remainder of the operand remains unchanged."

The 6 AND immediate operands are below:

![andimm](https://user-images.githubusercontent.com/11260919/32392454-333b3c36-c0ac-11e7-8fbb-1a5221b039cc.PNG)

The problem is in the fact that the remainder of the operand remains unchanged. So we can't use NILL because that would mean bits 0-47 would remain unchanged when they should be explicitly zeroed out. In fact we would probably need to generate a NIHF and NILF instruction because we need to make changes to the entire top half and bottom half of the register. 

I think a good way to think about this is that we are better off generating a RISBG if there's a 0 in the top half AND the bottom half of the constant (given that we have the leadingOnes or leadingZeros pattern). A 0 means that we need to explicitly zero out a particular bit in the register value if we use these AND Immediate instructions. If the zero exists in both halves then we will need two AND Immediate instructions. 

What you're suggesting does work for the converse case though. If we had something like 1000111111...111 instead, then we could use NILL to zero out those three bits and keep the rest of the bits in the register unchanged. This is exactly why the `if` clause on line 2569 seems a lot less restrictive than the one here because it can handle that case.

Perhaps this also deserves a better comment?